src/tree/mod.rs

```diff
@@ -737,122 +747,70 @@ impl RatchetTree {
-        let node_count = NodeIndex::from(nodes.len());
-        let size = node_count;
-        for i in 0..node_count.as_usize() {
-            let node_option = &nodes[i];
-            if let Some(node) = node_option {
-                match node.node_type {
-                    NodeType::Parent => {
-                        let left_index = treemath::left(NodeIndex::from(i))
-                            .expect("verify_integrity: Error when computing left child of node.");
-                        let right_index = treemath::right(NodeIndex::from(i), size.into())
-                            .expect("verify_integrity: Error when computing right child of node.");
-                        if right_index >= node_count {
-                            return false;
-                        }
-                        let left_option = &nodes[left_index.as_usize()];
-                        let right_option = &nodes[right_index.as_usize()];
-                        let own_hash = node.hash(ciphersuite).unwrap();
-                        if let Some(right) = right_option {
-                            if let Some(left) = left_option {
-                                let left_parent_hash = left.parent_hash().unwrap_or_else(Vec::new);
-                                let right_parent_hash =
-                                    right.parent_hash().unwrap_or_else(Vec::new);
-                                if (left_parent_hash != own_hash) && (right_parent_hash != own_hash)
-                                {
-                                    return false;
-                                }
-                                if left_parent_hash == right_parent_hash {
-                                    return false;
-                                }
-                            } else if right.parent_hash().unwrap() != own_hash {
-                                return false;
-                            }
-                        } else if let Some(left) = left_option {
-                            if left.parent_hash().unwrap() != own_hash {
-                                return false;
-                            }
-                        }
-                    }
-                    NodeType::Leaf => {
-                        if let Some(kp) = &node.key_package {
-                            if i % 2 != 0 {
-                                return false;
-                            }
-                            if kp.verify().is_err() {
-                                return false;
-                            }
-                        }
-                    }
 
-                    NodeType::Default => {}
-                }
+    /// Verify the parent hashes of the tree nodes. Returns `true` if all parent
+    /// hashes have successfully been verified and `false` otherwise.
+    pub fn verify_parent_hashes(&self) -> bool {
+        self.nodes.iter().enumerate().all(|(index, node)| {
+            if index % 2 == 1 && node.is_full_parent() {
```

Comment:
Just a nit, but I think casting to a `NodeIndex` and using `is_parent` should make it a bit clearer what the condition is.

Reply:
Good point, I replaced the modulo arithmetic everywhere.